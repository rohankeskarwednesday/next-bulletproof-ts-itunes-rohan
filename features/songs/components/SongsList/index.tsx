/**
 *
 * RepoList
 *
 */

import React from "react";
import get from "lodash/get";
import { Skeleton, Card, Alert } from "antd";
import { T, If } from "@common";
import { useIntl } from "react-intl";

interface SongListProps {
  data?: any;
  loading: boolean;
  error?: any;
}

const { Meta } = Card;

const SongsList: React.FC<SongListProps> = props => {
  const { data, loading, error } = props;
  const intl = useIntl();

  const items: any[] = get(data, "results", []);
  const totalCount: number = get(data, "resultCount", 0);
  const BlockText = props => <T display="block" {...props} />;

  const errorOccured = error != undefined && error.message != "Invalid params";

  return errorOccured ? (
    <Alert
      type="error"
      message={intl.formatMessage({ id: "request_failed" })}
      // message={}
      style={{ marginTop: "10px" }}
    />
  ) : (
    <If condition={items.length !== 0 || loading}>
      {totalCount !== 0 && (
        <BlockText id="matching_songs" values={{ totalCount }} style={{ marginTop: "10px" }} />
      )}
      <div data-testid="songs-list" style={{ display: "flex", flexWrap: "wrap" }}>
        <Skeleton loading={loading} active>
          {items.map((item, index: number) => (
            <Card
              key={index}
              hoverable
              style={{ width: 200, marginLeft: "20px", marginTop: "30px" }}
              cover={<img src={item.artworkUrl100} />}
            >
              <Meta
                title={
                  <BlockText
                    id="song_name"
                    values={{ collectionCensoredName: item.collectionCensoredName }}
                  />
                }
                description={""}
              />
            </Card>
          ))}
        </Skeleton>
      </div>
    </If>
  );
};

export default SongsList;
