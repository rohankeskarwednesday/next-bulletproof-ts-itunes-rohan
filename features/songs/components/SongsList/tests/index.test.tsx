// import { render } from "@testing-library/react";
import { fireEvent, render } from "@utils/testUtils";

import SongsList from "../index";

describe("<SongsList />", () => {
  const songsListProps = {
    loading: false,
    error: undefined,
    data: {
      resultCount: 1,
      results: [
        {
          wrapperType: "track",
          kind: "song",
          artistId: 258535972,
          collectionId: 258615649,
          trackId: 258618600,
          artistName: "Little Dragon",
          collectionName: "Little Dragon",
          trackName: "Test",
          collectionCensoredName: "Little Dragon",
          trackCensoredName: "Test",
          artistViewUrl: "https://music.apple.com/us/artist/little-dragon/258535972?uo=4",
          collectionViewUrl: "https://music.apple.com/us/album/test/258615649?i=258618600&uo=4",
          trackViewUrl: "https://music.apple.com/us/album/test/258615649?i=258618600&uo=4",
          previewUrl:
            "https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview115/v4/af/83/79/af837916-90f9-d503-bf08-13ffb9957f21/mzaf_16662673119862695322.plus.aac.p.m4a",
          artworkUrl100:
            "https://is4-ssl.mzstatic.com/image/thumb/Music/a0/74/81/mzi.ybszkvnq.tif/100x100bb.jpg",
        },
        {
          wrapperType: "track",
          kind: "song",
          artistId: 420203509,
          collectionId: 1440858222,
          trackId: 1440858321,
          artistName: "Bastille",
          collectionName: "Bad Blood (Bonus Track Version)",
          trackName: "Pompeii",
          collectionCensoredName: "Bad Blood (Bonus Track Version)",
          trackCensoredName: "Pompeii",
          artistViewUrl: "https://music.apple.com/us/artist/bastille/420203509?uo=4",
          collectionViewUrl:
            "https://music.apple.com/us/album/pompeii/1440858222?i=1440858321&uo=4",
          trackViewUrl: "https://music.apple.com/us/album/pompeii/1440858222?i=1440858321&uo=4",
          previewUrl:
            "https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview115/v4/1e/e3/20/1ee32037-f1a3-e46b-011f-76552e7c5fa2/mzaf_6001471003681391544.plus.aac.p.m4a",

          artworkUrl100:
            "https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/e6/d8/61/e6d86177-ae8c-e84e-dfcc-0042687066ed/13UAAIM41955.rgb.jpg/100x100bb.jpg",
        },
      ],
    },
  };

  it("should render and match the snapshot", () => {
    const { baseElement } = render(<SongsList {...songsListProps} />);
    expect(baseElement).toMatchSnapshot();
  });

  it("should contain 1 Song card", () => {
    const { getAllByTestId } = render(<SongsList {...songsListProps} />);
    expect(getAllByTestId("songs-list").length).toBe(1);
  });

  it("should play one song at a time", async () => {
    const { getByTestId } = render(<SongsList {...songsListProps} />);
    const firstTrack: any = getByTestId("258618600");
    const secondTrack = getByTestId("1440858321");

    fireEvent.play(firstTrack);
    fireEvent.play(secondTrack);

    expect(firstTrack.paused).toBe(true);
  });
});

window.HTMLMediaElement.prototype.load = () => {
  /* do nothing */
};

window.HTMLMediaElement.prototype.pause = () => {
  /* do nothing */
};
