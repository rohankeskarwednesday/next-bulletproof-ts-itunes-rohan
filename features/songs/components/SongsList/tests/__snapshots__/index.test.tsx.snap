// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SongsList /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="SongList"
      data-testid="songs-list"
    >
      <div
        class="ant-card ant-card-bordered ant-card-hoverable"
        style="width: 200px; margin-top: 30px; background-color: rgb(245, 245, 245);"
      >
        <div
          class="ant-card-cover"
        >
          <img
            src="https://is4-ssl.mzstatic.com/image/thumb/Music/a0/74/81/mzi.ybszkvnq.tif/100x100bb.jpg"
          />
        </div>
        <div
          class="ant-card-body"
        >
          <div
            class="ant-card-meta"
          >
            <div
              class="ant-card-meta-detail"
            >
              <div
                class="ant-card-meta-title"
              >
                <p
                  class="T__StyledText-sc-apdcd0-0 fjJkvM"
                  data-testid="t"
                  display="block"
                >
                  Little Dragon
                </p>
              </div>
            </div>
          </div>
          <div
            class="AudioPlayerContainer"
          >
            <audio
              controls=""
              src="https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview115/v4/af/83/79/af837916-90f9-d503-bf08-13ffb9957f21/mzaf_16662673119862695322.plus.aac.p.m4a"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
